libcalico-go/lib/clientv3/tier_e2e_test.go

```diff
@@ -0,0 +1,763 @@
+	. "github.com/onsi/ginkgo/extensions/table"
+	. "github.com/onsi/gomega"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	apiv3 "github.com/projectcalico/api/pkg/apis/projectcalico/v3"
+
+	"github.com/projectcalico/calico/libcalico-go/lib/apiconfig"
+	"github.com/projectcalico/calico/libcalico-go/lib/backend"
+	"github.com/projectcalico/calico/libcalico-go/lib/clientv3"
+	"github.com/projectcalico/calico/libcalico-go/lib/options"
+	"github.com/projectcalico/calico/libcalico-go/lib/testutils"
+	"github.com/projectcalico/calico/libcalico-go/lib/watch"
+)
+
+var _ = testutils.E2eDatastoreDescribe("Tier tests", testutils.DatastoreAll, func(config apiconfig.CalicoAPIConfig) {
+	ctx := context.Background()
+	defaultOrder := float64(1_000_000)
+	order1 := 99.999
+	order2 := 22.222
+	name1 := "t-1"
+	name2 := "t-12"
+	defaultName := "default"
+	namespace1 := "namespace-1"
+	spec1 := apiv3.TierSpec{
+		Order: &order1,
+	}
+	spec2 := apiv3.TierSpec{
+		Order: &order2,
+	}
+	defaultSpec := apiv3.TierSpec{
+		Order: &defaultOrder,
+	}
+
+	npName1 := name1 + ".networkp-1"
+	npSpec1 := apiv3.NetworkPolicySpec{
+		Tier:     name1,
+		Order:    &order1,
+		Ingress:  []apiv3.Rule{testutils.InRule1, testutils.InRule2},
+		Egress:   []apiv3.Rule{testutils.EgressRule1, testutils.EgressRule2},
+		Selector: "thing == 'value'",
+	}
+
+	gnpName1 := name1 + ".globalnetworkp-1"
+	gnpSpec1 := apiv3.GlobalNetworkPolicySpec{
+		Tier:           name1,
+		Order:          &order2,
+		Ingress:        []apiv3.Rule{testutils.InRule2, testutils.InRule1},
+		Egress:         []apiv3.Rule{testutils.EgressRule2, testutils.EgressRule1},
+		Selector:       "thing2 == 'value2'",
+		DoNotTrack:     true,
+		ApplyOnForward: true,
+	}
+
+	gnpName2 := name2 + ".globalnetworkp-1"
+	gnpSpec2 := apiv3.GlobalNetworkPolicySpec{
+		Tier:           name2,
+		Order:          &order2,
+		Ingress:        []apiv3.Rule{testutils.InRule2, testutils.InRule1},
+		Egress:         []apiv3.Rule{testutils.EgressRule2, testutils.EgressRule1},
+		Selector:       "thing2 == 'value2'",
+		DoNotTrack:     true,
+		ApplyOnForward: true,
+	}
+
+	DescribeTable("Tier e2e CRUD tests",
+		func(name1, name2, npName1, gnpName1, gnpName2, namespace1 string, spec1, spec2 apiv3.TierSpec, npSpec1 apiv3.NetworkPolicySpec, gnpSpec1, gnpSpec2 apiv3.GlobalNetworkPolicySpec) {
+			c, err := clientv3.New(config)
+			Expect(err).NotTo(HaveOccurred())
+
+			be, err := backend.NewClient(config)
+			Expect(err).NotTo(HaveOccurred())
+			be.Clean()
+
+			err = c.EnsureInitialized(ctx, "", "")
+			Expect(err).NotTo(HaveOccurred())
+
+			By("Creating the default tier with nil order")
+			res, outError := c.Tiers().Create(ctx, &apiv3.Tier{
+				ObjectMeta: metav1.ObjectMeta{Name: defaultName},
+				Spec:       apiv3.TierSpec{},
+			}, options.SetOptions{})
+			Expect(res).To(BeNil())
+			Expect(outError).To(HaveOccurred())
+			Expect(outError.Error()).To(Equal("operation Create is not supported on default: Invalid default Tier Order"))
```

Comment:
It probably doesn't matter, since our code should be creating the default tier always, but seems like "Create is not supported on default" is inconsistent with "Invalid default Tier order"

Do we support creating the default tier with this client? If so, the error should be "Invalid default Tier Order. The default tier must have order 1,000,000"

If we don't support it, then it should just say "Create is not supported on default"

Reply:
I'll fix this in another PR by moving this check to validator.